moneyed/classes.py

```diff
@@ -26,49 +28,61 @@ class Currency:
     canonical name, and countries the currency is used in.
     """
 
-    def __init__(self, code="", numeric=None, sub_unit=1, name=None, countries=None):
-        self.code = code
-        self.numeric = numeric
-        self.sub_unit = sub_unit
-        if name is not None:
-            self.name = name
-        if countries is not None:
-            self.countries = countries
-
-    def __hash__(self):
+    def __init__(
+        self,
+        code: str = "",
+        numeric: Optional[str] = None,
+        sub_unit: int = 1,
+        name: Optional[str] = None,
+        countries: Optional[List[str]] = None,
+    ) -> None:
+        self.code: Final = code
+        self.numeric: Final = numeric
+        self.sub_unit: Final = sub_unit
+        self._name: Final = name
+        self._countries: Final = countries
+
+    def __hash__(self) -> int:
         return hash(self.code)
 
-    def __eq__(self, other):
-        return type(self) is type(other) and self.code == other.code
+    def __eq__(self, other: object) -> bool:
+        return (
+            type(self) is type(other)
+            and self.code == other.code  # type: ignore[attr-defined]
+        )
 
-    def __ne__(self, other):
+    def __ne__(self, other: object) -> bool:
         return not self.__eq__(other)
 
-    def __repr__(self):
+    def __repr__(self) -> str:
         return self.code
 
-    def __lt__(self, other):
+    def __lt__(self, other: "Currency") -> bool:
         return self.code < other.code
 
-    def __le__(self, other):
+    def __le__(self, other: "Currency") -> bool:
         return self.code <= other.code
 
     @cached_property
-    def name(self):
+    def name(self) -> str:
         """
         Name of the currency in US locale. For backwards compat.
 
         Consider using get_name() instead, or babel.numbers.get_currency_name()
         """
-        return self.get_name("en_US")
+        return self._name if self._name is not None else self.get_name("en_US")
```

Comment:
I am inclined to have conditions as different statements, so it is more visible what is covered by the test suite in the coverage report.

Reply:
@Stranger6667 That makes sense. Btw, the reason I made this change is that mypy wasn't on board with overwriting the `name` method with a constant value.

Reply:
Yeah, the change seems very reasonable to me :)